File: Python3/201 - 400/229 Majority Element II.py
# 44ms, 99.80%
class Solution:
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        num1, num2, count1, count2, limit = 1, 0, 0, 0, len(nums) // 3
        for n in nums:
            if n == num1:
                count1 += 1
            elif n == num2:
                count2 += 1
            elif count1 == 0:
                num1, count1 = n, 1
            elif count2 == 0:
                num2, count2 = n, 1
            else:
                count1, count2 = count1 - 1, count2 - 1
        return [n for n in (num1, num2) if nums.count(n) > limit]

File: 201 - 400/229 Majority Element II.cc
// 8ms, 99.26%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();
class Solution {
public:
    vector<int> majorityElement(const vector<int> &nums) {
        vector<int> res;
        if (!nums.empty()) {
            int num[2] = {0, 1}, count[2] = {0, 0}, limit = nums.size() / 3;
            for (const auto n : nums)
                if (n == num[0]) {
                    ++count[0];
                } else if (n == num[1]) {
                    ++count[1];
                } else if (count[0] == 0) {
                    num[0] = n, ++count[0];
                } else if (count[1] == 0) {
                    num[1] = n, ++count[1];
                } else {
                    --count[0], --count[1];
                }

            if (num[0] > num[1]) swap(num[0], num[1]);
            count[0] = count[1] = 0;
            for (const auto n : nums)
                if (n == num[0]) {
                    ++count[0];
                } else if (n == num[1]) {
                    ++count[1];
                }

            if (count[0] > limit) res.push_back(num[0]);
            if (count[1] > limit) res.push_back(num[1]);
        }
        return res;
    }
};
